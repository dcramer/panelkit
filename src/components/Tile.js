import React, { Component } from "react";
import PropTypes from "prop-types";

import Icon from "./Icon";

export const TileProps = Object.freeze({
  hass: PropTypes.object.isRequired,
  name: PropTypes.string,
  entityId: PropTypes.string,
  icon: PropTypes.string,
  icons: PropTypes.objectOf(PropTypes.string),
  cameraList: PropTypes.array.isRequired,
});

const toTitleCase = (str) => {
  return str.replace(
    /\w\S*/g,
    (txt) => txt.charAt(0).toUpperCase() + txt.substr(1).toLowerCase()
  );
};

export default class Tile extends Component {
  static propTypes = TileProps;

  constructor(...params) {
    super(...params);
    this.state = this.getInitialState();
    this._activeSubscriptions = [];
  }

  getInitialState() {
    return {};
  }

  componentDidMount() {
    this.getWatchedEntityIds().forEach((entityId) => {
      this._activeSubscriptions.push(
        this.props.hass.subscribe(entityId, this.onStateChange)
      );
    });
  }

  componentWillUnmount() {
    this._activeSubscriptions.forEach((sub) => {
      this.props.hass.unsubscribe(...sub);
    });
  }

  /* Defines a list of entity IDs to monitor for update.
   *
   * When an entity is updated, it will call `onStateChange`, which
   * by default will simply force the tile to re-render.
   */
  getWatchedEntityIds() {
    return [];
  }

  getEntity(entityId) {
    return this.props.hass.getState(entityId);
  }

  callService(domain, service, serviceData) {
    return this.props.hass.callService(domain, service, serviceData);
  }

  onStateChange = (_entityId, _newState, _oldState) => {
    // XXX(dcramer): Yes, you shouldn't do this. No I don't care about your opinions.
    this.forceUpdate();
  };

  getIcon() {
    if (this.props.entityId) {
      return this.getIconForEntity(this.props.entityId);
    }
    return this.props.icon;
  }

  getIconForEntity(entityId) {
    const { state } = this.getEntity(entityId);
    if (this.props.icons && this.props.icons[state])
      return this.props.icons[state];
    return this.props.icon;
  }

  onClick = null;

  getClassNames() {
    if (!this.props.entityId) return "";
    const { state } = this.getEntity(this.props.entityId);
    return `state-${state}`;
  }

  render() {
    const status = this.renderStatus();
    const label = this.renderLabel();
    const cover = this.renderCover();
    return (
      <div
        className={`tile ${this.getClassNames()} ${
          this.onClick && "clickable"
        }`}
        onClick={this.onClick}
        style={{ cursor: this.onClick ? "pointer" : "normal" }}
      >
        <div className="tile-container">
          {cover && <div className="tile-cover">{cover}</div>}
          {status && <div className="tile-status">{status}</div>}
          <div className="tile-body">{this.renderBody()}</div>
          {label && <div className="tile-label">{label}</div>}
        </div>
      </div>
    );
  }

  renderBody() {
    const icon = this.getIcon();
    if (!icon) return null;
    return (
      <span className="tile-icon">
        <Icon name={icon} />
      </span>
    );
  }

  renderCover() {}

  renderLabel() {
    if (!this.props.entityId) return null;
    const {
      attributes: { friendly_name },
    } = this.getEntity(this.props.entityId);
    return this.props.name || friendly_name;
  }

  renderStatus() {
    if (!this.props.entityId) return null;
    const { state } = this.getEntity(this.props.entityId);
    return toTitleCase(state);
  }
}
