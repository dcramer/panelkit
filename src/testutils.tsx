import React from "react";
import { render } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

import { Entity } from "./types";
import HomeAssistant, { SuggestedChanges } from "./hass";

export const URL = "http://localhost:8123";

export const ACCESS_TOKEN = "example.access.token";

export const DEFAULT_CONFIG = {
  url: URL,
  accessToken: ACCESS_TOKEN,
  tiles: [],
};

export const simulateTouch = (container: any) => {
  userEvent.click(container);
};

export const getTileProps = (entity: Entity) => {
  const hass = new HomeAssistant(DEFAULT_CONFIG);

  hass.callService = jest
    .fn()
    .mockImplementation(
      (
        domain: string,
        service: string,
        serviceData: any,
        suggestedChanges: SuggestedChanges | null = null
      ) => {
        return Promise.resolve();
      }
    );
  hass.sendMessage = jest.fn().mockImplementation(() => {
    return Promise.resolve();
  });

  hass.populateEntityCache([entity]);

  return {
    entityId: entity.entity_id,
    hass,
    cameraList: [],
  };
};

export const renderTile = (
  Component: any,
  entity: Entity,
  props: { [key: string]: any } = {}
) => {
  const tileProps = {
    ...getTileProps(entity),
    ...props,
  };

  const result = render(<Component {...tileProps} />);

  return {
    ...result,
    hass: tileProps.hass,
    container: result.container.firstChild,
  };
};
