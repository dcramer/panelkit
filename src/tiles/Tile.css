.tile {
  position: relative;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  overflow: hidden;
  width: 100%;
  height: 100%;
  background: var(--tile-bg-color);
  -webkit-column-break-inside: avoid;
  box-sizing: border-box;
  box-shadow: 1px 1px 0 rgba(54, 47, 36, 0.35),
    -1px -1px 0 rgba(200, 200, 200, 0.1);
  padding: 6px;
}

.tile.touchable:focus,
.tile.touchable:hover {
  border: 3px solid #adb8bc;
  padding: 3px;
}

.tile.touchable::after {
  content: " ";
  display: block;
  left: 5px;
  top: 5px;
  height: 6px;
  width: 6px;
  border-left: 3px solid #adb8bc;
  border-top: 3px solid #adb8bc;
  opacity: 0.5;
  position: absolute;
}

.tile.touchable:focus::after,
.tile.touchable:hover::after {
  left: 2px;
  top: 2px;
  opacity: 1;
}

.tile-container {
  position: relative;
  height: 100%;
  width: 100%;
}

.tile.state-off {
  color: var(--switch-off-color);
}

.tile.state-off .tile-icon {
  opacity: 0.5;
}

.tile.state-on {
  color: var(--switch-on-color);
}

.tile.state-armed_home,
.tile.state-armed_night,
.tile.state-armed_away {
  color: var(--alarm-armed-color);
}

.tile.state-disarmed {
  color: var(--alarm-disarmed-color);
}

.tile-body {
  height: 100%;
  width: 100%;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  font-size: 32pt;
  line-height: 32pt;
  text-transform: uppercase;
}

.tile-body .unit {
  font-weight: 400;
  font-size: 0.4em;
  vertical-align: super;
  margin-top: -15%;
  margin-left: 5px;
}

.tile-status .unit {
  font-weight: 400;
  font-size: 0.9em;
  margin-left: 5px;
}

.tile-icon {
  width: 50px;
  height: 50px;
}

.tile-icon svg {
  width: 100%;
  height: 100%;
}

.tile-cover {
  position: absolute;
  left: -6px;
  right: -6px;
  top: -6px;
  bottom: -6px;
}
.tile-cover img {
  width: 100%;
  height: 100%;
  object-fit: cover;
  object-position: -50% 50%;
}

.tile-title,
.tile-subtitle,
.tile-status {
  position: absolute;
  padding: 1px 0;
  max-width: 100%;
  overflow: hidden;
  white-space: nowrap;
  text-overflow: ellipsis;
  background: rgba(0, 0, 0, 0.1);
  max-width: 100%;
}

.tile-title {
  font-weight: bold;
  color: var(--text-color);
  left: 0;
  bottom: 0;
}

.tile-subtitle {
  font-weight: 400;
  color: var(--text-color);
  left: 0;
  bottom: 20px;
  opacity: 0.9;
}

.tile-status {
  font-weight: bold;
  text-align: right;
  top: 0;
  right: 0;
  font-size: 0.9em;
}
