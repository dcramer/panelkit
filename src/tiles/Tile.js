import React, { Component } from "react";
import PropTypes from "prop-types";
import { toast } from "react-toastify";

import "./Tile.css";
import Icon from "../components/Icon";
import { toTitleCase } from "../utils";

const LONG_PRESS_TIME = 300;

export const TileConfig = Object.freeze({
  title: PropTypes.string,
  subtitle: PropTypes.string,
  entityId: PropTypes.string,
  icon: PropTypes.string,
  icons: PropTypes.objectOf(PropTypes.string),
});

export const TileProps = Object.freeze({
  ...TileConfig,
  hass: PropTypes.object.isRequired,
  cameraList: PropTypes.array.isRequired,
});

export default class Tile extends Component {
  static propTypes = TileProps;

  static defaultIcon;

  constructor(...params) {
    super(...params);
    this.state = {
      modalIsOpen: false,
      ...this.getInitialState(),
    };
    this._activeSubscriptions = [];
    this._clickTimer = null;
  }

  getInitialState() {
    return {};
  }

  componentDidMount() {
    this.getWatchedEntityIds().forEach((entityId) => {
      this._activeSubscriptions.push(
        this.props.hass.subscribe(entityId, this.onStateChange)
      );
    });
  }

  componentWillUnmount() {
    this._activeSubscriptions.forEach((sub) => {
      this.props.hass.unsubscribe(...sub);
    });
    if (this._clickTimer) {
      clearTimeout(this._clickTimer);
      this._clickTimer = null;
    }
  }

  handleMouseDown = (e) => {
    e.preventDefault();
    if ((e.button === 0 && e.ctrlKey) || e.button > 0) return;
    return this.handleButtonPress();
  };

  handleTouchStart = (e) => {
    e.preventDefault();
    if (e.touches.length > 1) return;
    if ((e.button === 0 && e.ctrlKey) || e.button > 0) return;
    return this.handleButtonPress();
  };

  handleTouchMove = (e) => {
    // We handle the touch move event to avoid a "long press" event from being
    // fired after someone touchhes and attempts to scroll
    if (this._clickTimer) {
      clearTimeout(this._clickTimer);
      this._clickTimer = null;
    }
  };

  handleButtonPress = (e) => {
    // TODO(dcramer): i cant understand why onClick is firing outsie of the element scope
    if (this.state.modalIsOpen) return;
    // if we've got a long press handler define, start our countdown, which will get
    // interrupted on press/mouse release
    if (this.onLongTouch) {
      this._clickTimer = setTimeout(() => {
        this.onLongTouch();
        this._clickTimer = null;
      }, LONG_PRESS_TIME);
    }
  };

  handleButtonRelease = (e) => {
    if (this.state.modalIsOpen) return;
    e && e.preventDefault();
    if (this.onLongTouch && this._clickTimer) {
      clearTimeout(this._clickTimer);
      this._clickTimer = null;
      this.onTouch && this.onTouch();
    } else if (!this.onLongTouch) {
      this.onTouch && this.onTouch();
    }
  };

  handleButtonLeave = (e) => {
    e && e.preventDefault();
    if (this._clickTimer) {
      clearTimeout(this._clickTimer);
      this._clickTimer = null;
    }
  };

  onTouch = null;

  onLongTouch = null;

  /* Defines a list of entity IDs to monitor for update.
   *
   * When an entity is updated, it will call `onStateChange`, which
   * by default will simply force the tile to re-render.
   */
  getWatchedEntityIds() {
    if (this.props.entityId) return [this.props.entityId];
    return [];
  }

  getEntity(...params) {
    return this.props.hass.getEntity(...params);
  }

  callService(domain, service, ...params) {
    return this.props.hass
      .callService(domain, service, ...params)
      .catch((err) => {
        toast.error(
          `Error occured when calling ${domain}.${service}: ${err.message}`
        );
      });
  }

  onStateChange = (_entityId, _newState) => {
    // XXX(dcramer): Yes, you shouldn't do this. No I don't care about your opinions.
    this.forceUpdate();
  };

  getIcon() {
    if (this.props.entityId) {
      return this.getIconForEntity(this.props.entityId);
    }
    return this.props.icon || this.getDefaultIcon();
  }

  getIconForEntity(entityId) {
    const { state } = this.getEntity(entityId);
    if (this.props.icons && this.props.icons[state])
      return this.props.icons[state];
    return this.props.icon || this.getDefaultIcon();
  }

  /*
   * Return the default icon when no overrides are present.
   */
  getDefaultIcon() {
    return this.constructor.defaultIcon;
  }

  getClassNames() {
    if (!this.props.entityId) return "";
    const { state } = this.getEntity(this.props.entityId);
    return `state-${state}`;
  }

  isTouchable() {
    return this.onTouch || this.onLongTouch;
  }

  openModal = () => {
    if (this.state.modalIsOpen) return;
    this.setState({ modalIsOpen: true });
  };

  closeModal = () => {
    this.setState({ modalIsOpen: false });
  };

  render() {
    const isTouchable = this.isTouchable();
    const status = this.renderStatus();
    const title = this.renderTitle();
    const subtitle = this.renderSubtitle();
    const cover = this.renderCover();
    return (
      <div
        className={`tile ${this.getClassNames()} ${isTouchable && "touchable"}`}
        onMouseDown={this.handleMouseDown}
        onMouseUp={this.handleButtonRelease}
        onMouseLeave={this.handleButtonLeave}
        onTouchStart={this.handleTouchStart}
        onTouchEnd={this.handleButtonRelease}
        onTouchMove={this.handleTouchMove}
        style={{ cursor: isTouchable ? "pointer" : "normal" }}
      >
        <div className="tile-container">
          {cover && <div className="tile-cover">{cover}</div>}
          {status && <div className="tile-status">{status}</div>}
          <div className="tile-body">{this.renderBody()}</div>
          {title && <div className="tile-title">{title}</div>}
          {subtitle && <div className="tile-subtitle">{subtitle}</div>}
          {this.state.modalIsOpen &&
            this.renderModal({
              hass: this.props.hass,
              entityId: this.props.entityId,
              isOpen: this.state.modalIsOpen,
              onRequestClose: this.closeModal,
              cameraList: this.props.cameraList,
            })}
        </div>
      </div>
    );
  }

  renderModal() {}

  renderBody() {
    const icon = this.getIcon();
    if (!icon) return null;
    return (
      <span className="tile-icon">
        <Icon name={icon} />
      </span>
    );
  }

  renderCover() {}

  renderSubtitle() {
    return this.props.subtitle;
  }

  renderTitle() {
    if (!this.props.entityId) return null;
    const {
      attributes: { friendly_name },
    } = this.getEntity(this.props.entityId);
    return this.props.title || friendly_name;
  }

  renderStatus() {
    if (!this.props.entityId) return null;
    const { state } = this.getEntity(this.props.entityId);
    return toTitleCase(state);
  }
}
