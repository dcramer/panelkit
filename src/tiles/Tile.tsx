import React, { Component } from "react";
import { toast } from "react-toastify";
import * as Sentry from "@sentry/browser";

import "./Tile.css";
import LoadingIndicator from "../components/LoadingIndicator";
import Icon from "../components/Icon";
import { startTransaction } from "../sentry";
import { toTitleCase } from "../utils";

import HomeAssistant from "../hass";
import { Entity, TileComponentConfig } from "../types";

const LONG_PRESS_TIME = 1000;

// maximum time to wait for a state change
// after a service call (controls isLoading indicator)
const SERVICE_CALL_TIMEOUT = 30000;

export interface TileProps extends TileComponentConfig {
  hass: HomeAssistant;
  cameraList: string[];
}

export interface TileState {
  modalIsOpen: boolean;
  isLoading: boolean;
}

export type ModalParams = {
  entityId?: string;
  hass: HomeAssistant;
  callService: Function;
  cameraList: string[];
  isOpen: boolean;
  onRequestClose: () => void;
};

export default class Tile<
  T extends TileProps = TileProps,
  S extends TileState = TileState
> extends Component<T, S> {
  public static defaultIcon: string;

  private _activeSubscriptions: string[];
  private _longPressTimer: number | null;
  private _loadingTimer: number | null;
  private _wasScrolling: boolean;

  state: S;

  constructor(props: T, context?: any) {
    super(props, context);
    this._activeSubscriptions = [];
    this._longPressTimer = null;
    this._loadingTimer = null;
    this._wasScrolling = false;

    this.callService = this.callService.bind(this);

    this.state = this.getInitialState();
  }

  getInitialState(): any {
    return { modalIsOpen: false, isLoading: false };
  }

  componentDidMount() {
    this.getWatchedEntityIds().forEach((entityId) => {
      this._activeSubscriptions.push(
        this.props.hass.subscribe(entityId, this.onStateChange)
      );
    });
  }

  componentWillUnmount() {
    this._activeSubscriptions.forEach((sub) => {
      this.props.hass.unsubscribe(sub);
    });
    if (this._longPressTimer) {
      clearTimeout(this._longPressTimer);
      this._longPressTimer = null;
    }
    if (this._loadingTimer) clearTimeout(this._loadingTimer);
  }

  handleMouseDown = async (e: React.MouseEvent<HTMLElement>) => {
    if ((e.button === 0 && e.ctrlKey) || e.button > 0) {
      console.debug("[panelkit] Ignoring click due to mouse button(s)");
      return;
    }
    return this.handleButtonPress(e);
  };

  handleTouchStart = async (e: React.TouchEvent<HTMLElement>) => {
    if (e.touches.length > 1) return;
    return this.handleButtonPress(e);
  };

  handleTouchMove = async (e: any) => {
    // We handle the touch move event to avoid a "long press" event from being
    // fired after someone touchhes and attempts to scroll
    if (this._longPressTimer) {
      clearTimeout(this._longPressTimer);
      this._longPressTimer = null;
    }
    this._wasScrolling = true;
  };

  handleButtonPress = async (e: any) => {
    if (this.onLongTouch) {
      this._longPressTimer = setTimeout(async () => {
        this.handleOnLongTouch();
        this._longPressTimer = null;
      }, LONG_PRESS_TIME);
    }
  };

  handleButtonRelease = async (
    e: React.TouchEvent<HTMLElement> | React.MouseEvent<HTMLElement>
  ) => {
    e.cancelable && e.preventDefault();
    if (this._wasScrolling) {
      console.debug("[panelkit] Ignoring click due to scrolling");
      this._wasScrolling = false;
      return;
    }
    if (this.onLongTouch && this._longPressTimer) {
      clearTimeout(this._longPressTimer);
      this._longPressTimer = null;
      this.handleOnTouch();
    } else if (!this.onLongTouch && this.onTouch) {
      this.handleOnTouch();
    }
  };

  handleButtonLeave = async (
    e: React.TouchEvent<HTMLElement> | React.MouseEvent<HTMLElement>
  ) => {
    if (this._longPressTimer) {
      clearTimeout(this._longPressTimer);
      this._longPressTimer = null;
    }
  };

  handleOnTouch = async () => {
    if (!this.onTouch) return;
    Sentry.withScope(async (scope) => {
      scope.setTransaction(`panelkit.touch ${this.constructor.name}`);
      const transaction = startTransaction({
        name: `panelkit.touch ${this.constructor.name}`,
        op: "tile.touch",
        description: this.constructor.name,
      });
      this.onTouch && (await this.onTouch());
      if (transaction) transaction.finish();
    });
  };

  handleOnLongTouch = async () => {
    if (!this.onLongTouch) return;
    Sentry.withScope(async (scope) => {
      scope.setTransaction(`panelkit.long-touch ${this.constructor.name}`);
      const transaction = startTransaction({
        name: `panelkit.long-touch ${this.constructor.name}`,
        op: "tile.long-touch",
        description: this.constructor.name,
      });
      this.onLongTouch && (await this.onLongTouch());
      if (transaction) transaction.finish();
    });
  };

  onTouch: (() => void) | null = null;

  onLongTouch: (() => void) | null = null;

  /* Defines a list of entity IDs to monitor for update.
   *
   * When an entity is updated, it will call `onStateChange`, which
   * by default will simply force the tile to re-render.
   */
  getWatchedEntityIds(): string[] {
    const { entityId } = this.props;
    if (entityId) return [entityId as string];
    return [];
  }

  getEntity(entityId: string): Entity {
    return this.props.hass.getEntity(entityId);
  }

  async callService(
    domain: string,
    service: string,
    serviceData: any,
    suggestedChanges?: any
  ) {
    if (
      this.serviceCallShouldWaitOnState(
        domain,
        service,
        serviceData,
        suggestedChanges
      )
    ) {
      this.setState({ isLoading: true });
      this._loadingTimer = setTimeout(() => {
        this.setState({ isLoading: false });
      }, SERVICE_CALL_TIMEOUT);
    }

    return this.props.hass
      .callService(domain, service, serviceData, suggestedChanges)
      .catch((err: Error) => {
        toast.error(
          `Error occured when calling ${domain}.${service}: ${err.message}`
        );
        this.setState({ isLoading: false });
      });
  }

  serviceCallShouldWaitOnState(
    domain: string,
    service: string,
    serviceData: any,
    suggestedChanges?: any
  ) {
    switch (domain) {
      case "scene":
        return false;
      default:
        return true;
    }
  }

  onStateChange = (entityId: string, newState: Entity) => {
    this.setState({ isLoading: false });
  };

  getState(state: string): string {
    if (this.props.states && this.props.states[state])
      return this.props.states[state] as string;
    return toTitleCase(state);
  }

  getIcon(): string {
    const { entityId, icon } = this.props;
    if (entityId) return this.getIconForEntity(entityId as string);
    if (icon) return icon as string;
    return this.getDefaultIcon();
  }

  getIconForEntity(entityId: string): string {
    const { state } = this.getEntity(entityId);
    if (this.props.icons && this.props.icons[state])
      return this.props.icons[state] as string;
    if (this.props.icon) return this.props.icon as string;
    return this.getDefaultIcon();
  }

  /*
   * Return the default icon when no overrides are present.
   */
  getDefaultIcon(): string {
    if (this.props.entityId) {
      const {
        attributes: { icon },
      } = this.getEntity(this.props.entityId as string);
      if (icon) return icon.split("mdi:", 2)[1];
    }
    return (this.constructor as typeof Tile).defaultIcon;
  }

  getClassNames(): string {
    const { entityId } = this.props;
    if (!entityId) return "";
    const { state } = this.getEntity(entityId as string);
    return `state-${state}`;
  }

  isTouchable(): boolean {
    return !!(this.onTouch || this.onLongTouch);
  }

  openModal = () => {
    if (this.state.modalIsOpen) return;
    this.setState({ modalIsOpen: true });
  };

  closeModal = () => {
    this.setState({ modalIsOpen: false });
  };

  render() {
    const { isLoading } = this.state;
    const isTouchable = this.isTouchable();
    const status = this.renderStatus();
    const title = this.renderTitle();
    const subtitle = this.renderSubtitle();
    const cover = this.renderCover();
    return (
      <div
        className={`tile ${this.getClassNames()} ${isTouchable && "touchable"}`}
        onMouseDown={this.handleMouseDown}
        onMouseUp={this.handleButtonRelease}
        onMouseLeave={this.handleButtonLeave}
        onTouchStart={this.handleTouchStart}
        onTouchEnd={this.handleButtonRelease}
        onTouchMove={this.handleTouchMove}
        style={{ cursor: isTouchable ? "pointer" : "normal" }}
        data-testid="tile"
      >
        <div className="tile-container">
          {cover && <div className="tile-cover">{cover}</div>}
          {status && <div className="tile-status">{status}</div>}
          <div className="tile-body">{this.renderBody()}</div>
          {title && <div className="tile-title">{title}</div>}
          {subtitle && <div className="tile-subtitle">{subtitle}</div>}
          {isLoading && (
            <div className="tile-loading">
              <LoadingIndicator />
            </div>
          )}
          {this.state.modalIsOpen &&
            this.renderModal({
              hass: this.props.hass,
              entityId: this.props.entityId,
              isOpen: this.state.modalIsOpen,
              onRequestClose: this.closeModal,
              cameraList: this.props.cameraList,
              callService: this.callService,
            })}
        </div>
      </div>
    );
  }

  renderModal(params: ModalParams): any | null {
    return null;
  }

  renderBody() {
    const icon = this.getIcon();
    if (!icon) return null;
    return (
      <span className="tile-icon">
        <Icon name={icon as string} />
      </span>
    );
  }

  renderCover(): any | null {
    return null;
  }

  renderSubtitle(): string | null {
    const { subtitle } = this.props;
    if (!subtitle) return null;
    return subtitle as string;
  }

  renderTitle(): any | null {
    const { entityId } = this.props;
    if (!entityId) return null;
    const {
      attributes: { friendly_name },
    } = this.getEntity(entityId as string);
    return this.props.title || friendly_name || "";
  }

  renderStatus(): any | null {
    const { entityId } = this.props;
    if (!entityId) return null;
    const { state } = this.getEntity(entityId as string);
    if (!state) return null;
    return this.getState(state) || null;
  }
}
