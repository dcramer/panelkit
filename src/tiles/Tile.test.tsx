import { renderTile } from "../testutils";
import Tile from "./Tile";

describe("renderStatus", () => {
  test("uses states when available", async () => {
    const { container, getByText } = renderTile(
      Tile,
      {
        entity_id: "alarm_control_panel.test",
        state: "disarmed",
        attributes: {
          friendly_name: "Test Alarm",
        },
      },
      {
        states: {
          disarmed: "Really Disarmed",
        },
      }
    );

    expect(getByText(/Really Disarmed/i)).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  test("title cases state by default", async () => {
    const { container, getByText } = renderTile(
      Tile,
      {
        entity_id: "alarm_control_panel.test",
        state: "not_armed",
        attributes: {
          friendly_name: "Test Alarm",
        },
      },
      {
        states: {
          disarmed: "Really Disarmed",
        },
      }
    );

    expect(getByText(/Not armed/i)).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });
});
