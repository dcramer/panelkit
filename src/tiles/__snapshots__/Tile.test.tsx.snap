// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renderStatus uses states 1`] = `
<div
  class="tile state-disarmed false"
  data-testid="tile"
  style="cursor: normal;"
>
  <div
    class="tile-container"
  >
    <div
      class="tile-status"
    >
      Really Disarmed
    </div>
    <div
      class="tile-body"
    />
    <div
      class="tile-title"
    >
      Test Alarm
    </div>
  </div>
</div>
`;

exports[`renderStatus uses states when available 1`] = `
<div
  class="tile state-disarmed false"
  data-testid="tile"
  style="cursor: normal;"
>
  <div
    class="tile-container"
  >
    <div
      class="tile-status"
    >
      Really Disarmed
    </div>
    <div
      class="tile-body"
    />
    <div
      class="tile-title"
    >
      Test Alarm
    </div>
  </div>
</div>
`;
